event_model/__init__.py

```diff
@@ -25,6 +26,82 @@ class DocumentNames(Enum):
     resource = 'resource'
 
 
+class DocumentRouter:
+    """
+    Route each document by type to a corresponding method.
+
+    When an instance is called with a document type and a document like::
+
+        router(name, doc)
+
+    the document is passed to the method of the corresponding name, as in
+
+        getattr(router, name)(doc)
+
+    The method is expected to return a valid document of the same type. It may
+    be the original instance (passed through), a copy, or a different dict
+    altogether.
+
+    Finally, the call to ``router(name, doc))`` returns::
+
+        (name, getattr(router, name)(doc))
+    """
+    def __call__(self, name, doc):
+        """
+        Process a document.
+
+        Parameters
+        ----------
+        name : string
+        doc : dict
+
+        Returns
+        -------
+        name, new_doc : string, dict
+            The same name as what was passed in, and a doc that may be the same
+            instance as doc, a copy of doc, or a different dict altogether.
+        """
+        return name, getattr(self, name)(doc)
```

Comment:
Suggest changing to:

```suggestion
         ret = getattr(self, name)(doc)
         return name, ret if ret is not None else doc
```

so that if the down-stream implementer forgets to return it will not break everything.